File: FirebaseRepository.swift
// FirebaseRepository.swift — ParentGuard iOS
// All Firestore reads and writes.  Mirrors FirebaseRepository.kt on Android.
//
// The Firestore schema is shared across both platforms:
//   families/{familyId}/children/{childId}/
//     profile/info      — DeviceProfile
//     appUsage/{date_pkg} — AppUsageRecord
//     blockedApps/{pkg}  — BlockedApp (parent writes, child reads)
//     screenTimeLimits/limits — ScreenTimeLimit (parent writes, child reads)
//     alerts/{id}        — AlertRecord

import Foundation
import FirebaseAuth
import FirebaseFirestore

final class FirebaseRepository {

    static let shared = FirebaseRepository()

    private let db   = Firestore.firestore()
    private let auth = Auth.auth()
    private let prefs = AppPreferences.shared

    private init() {}

    // ── Auth guard ────────────────────────────────────────────────────────────

    /// Every write path calls this first. Throws if unauthenticated or IDs are blank.
    private func requireAuth() throws {
        guard auth.currentUser != nil else {
            throw PGError.unauthenticated
        }
        guard !prefs.familyId.isEmpty else { throw PGError.missingFamilyId }
        guard !prefs.childId.isEmpty  else { throw PGError.missingChildId  }
    }

    // ── Paths ─────────────────────────────────────────────────────────────────

    private var childRef: DocumentReference {
        db.collection("families").document(prefs.familyId)
          .collection("children").document(prefs.childId)
    }

    // ── Upload ────────────────────────────────────────────────────────────────

    /// Upload the device profile (heartbeat — called every 15 minutes).
    func uploadDeviceProfile(_ profile: DeviceProfile) async throws {
        try requireAuth()
        let data = try Firestore.Encoder().encode(profile)
        try await childRef.collection("profile").document("info")
            .setData(data, merge: true)
    }

    /// Upload app usage records. Batched in groups of 400 (Firestore limit is 500).
    func uploadAppUsage(_ records: [AppUsageRecord]) async throws {
        guard !records.isEmpty else { return }
        try requireAuth()

        for chunk in records.chunked(into: 400) {
            let batch = db.batch()
            for record in chunk {
                let docId = "\(record.date)_\(record.bundleIdentifier.replacing(".", with: "_"))"
                let ref   = childRef.collection("appUsage").document(docId)
                let data  = try Firestore.Encoder().encode(record)
                batch.setData(data, forDocument: ref, merge: true)
            }
            try await batch.commit()
        }
    }

    /// Upload an alert (app blocked, bedtime violation, etc.).
    func uploadAlert(_ alert: AlertRecord) async throws {
        try requireAuth()
        let data = try Firestore.Encoder().encode(alert)
        try await childRef.collection("alerts").document(alert.id)
            .setData(data, merge: true)
    }

    // ── Download (parent commands) ────────────────────────────────────────────

    /// Fetch the current blocked apps list from Firestore.
    func fetchBlockedApps() async throws -> [BlockedApp] {
        try requireAuth()
        let snapshot = try await childRef.collection("blockedApps").getDocuments()
        return try snapshot.documents.compactMap {
            try $0.data(as: BlockedApp.self)
        }
    }

    /// Fetch screen time limits.
    func fetchScreenTimeLimits() async throws -> ScreenTimeLimit {
        try requireAuth()
        let doc = try await childRef.collection("screenTimeLimits").document("limits").getDocument()
        return try doc.data(as: ScreenTimeLimit.self) ?? ScreenTimeLimit()
    }

    // ── Real-time listener (parent command push) ──────────────────────────────

    /// Listen for changes to blocked apps in real time. Calls the handler whenever
    /// the parent adds or removes a blocked app from the dashboard.
    @discardableResult
    func listenForBlockedAppsChanges(
        handler: @escaping ([BlockedApp]) -> Void
    ) -> ListenerRegistration? {
        guard (try? requireAuth()) != nil else { return nil }

        return childRef.collection("blockedApps")
            .addSnapshotListener { snapshot, error in
                guard let snapshot, error == nil else { return }
                let apps = (try? snapshot.documents.compactMap {
                    try $0.data(as: BlockedApp.self)
                }) ?? []
                handler(apps)
            }
    }

    @discardableResult
    func listenForScreenTimeLimitChanges(
        handler: @escaping (ScreenTimeLimit) -> Void
    ) -> ListenerRegistration? {
        guard (try? requireAuth()) != nil else { return nil }

        return childRef.collection("screenTimeLimits").document("limits")
            .addSnapshotListener { snapshot, error in
                guard let snapshot, error == nil else { return }
                if let limit = try? snapshot.data(as: ScreenTimeLimit.self) {
                    handler(limit)
                }
            }
    }

    // ── Errors ────────────────────────────────────────────────────────────────

    enum PGError: LocalizedError {
        case unauthenticated
        case missingFamilyId
        case missingChildId

        var errorDescription: String? {
            switch self {
            case .unauthenticated: return "No authenticated Firebase user — cannot upload."
            case .missingFamilyId: return "familyId is blank — complete setup first."
            case .missingChildId:  return "childId is blank — should not happen."
            }
        }
    }
}

// ─── Array chunking helper ────────────────────────────────────────────────────

extension Array {
    func chunked(into size: Int) -> [[Element]] {
        stride(from: 0, to: count, by: size).map {
            Array(self[$0 ..< Swift.min($0 + size, count)])
        }
    }
}

// ─── String.replacing helper (backport for < iOS 16.0 regex syntax) ──────────

extension String {
    func replacing(_ char: Character, with replacement: String) -> String {
        self.replacingOccurrences(of: String(char), with: replacement)
    }
}

File: FirebaseRepository.kt
package com.parentguard.monitor

import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import kotlinx.coroutines.tasks.await

/**
 * FirebaseRepository handles all Firestore read/write operations.
 *
 * Firestore structure:
 *   families/{familyId}/
 *     children/{childDeviceId}/
 *       profile        (DeviceProfile doc)
 *       sms/           (collection)
 *       calls/         (collection)
 *       appUsage/      (collection, one doc per date per app)
 *       blockedApps/   (collection)
 *       screenTimeLimits (doc)
 */
class FirebaseRepository {

    private val db   = FirebaseFirestore.getInstance()
    private val auth = FirebaseAuth.getInstance()

    private val prefs = AppPreferences.instance

    private val familyId: String get() = prefs.familyId
    private val childId:  String get() = prefs.childId

    /**
     * Security guard: every write path calls this first.
     * If the device has no authenticated Firebase user, writes are rejected locally —
     * Firestore security rules would reject them anyway, but failing fast avoids
     * wasting network and gives a clear log entry.
     */
    private fun requireAuth() {
        checkNotNull(auth.currentUser) {
            "FirebaseRepository: write attempted without an authenticated user — aborting"
        }
        require(familyId.isNotBlank()) { "familyId must not be blank" }
        require(childId.isNotBlank())  { "childId must not be blank" }
    }

    private fun childRef() = db
        .collection("families").document(familyId)
        .collection("children").document(childId)

    // ─── Batch helper ────────────────────────────────────────────────────────

    /**
     * Firestore batches are limited to 500 operations.
     * FIX: uploadSmsRecords / uploadCallRecords / uploadAppUsage previously created
     * a single batch over the entire list — crashing silently on first sync when
     * all historical records are fetched at once (potentially thousands of rows).
     * Now every upload function chunks into 400-item batches (safe margin below 500).
     */
    private suspend fun <T> batchWrite(
        items: List<T>,
        docRef: (T) -> com.google.firebase.firestore.DocumentReference,
        toMap: (T) -> Any = { it as Any }
    ) {
        items.chunked(400).forEach { chunk ->
            val batch = db.batch()
            chunk.forEach { item ->
                batch.set(docRef(item), toMap(item), SetOptions.merge())
            }
            batch.commit().await()
        }
    }

    // ─── Upload ──────────────────────────────────────────────────────────────

    suspend fun uploadSmsRecords(records: List<SmsRecord>) {
        if (records.isEmpty()) return
        requireAuth()
        batchWrite(records, { childRef().collection("sms").document(it.id) })
        Log.d(TAG, "Uploaded ${records.size} SMS records")
    }

    suspend fun uploadCallRecords(records: List<CallRecord>) {
        if (records.isEmpty()) return
        requireAuth()
        batchWrite(records, { childRef().collection("calls").document(it.id) })
        Log.d(TAG, "Uploaded ${records.size} call records")
    }

    suspend fun uploadAppUsage(records: List<AppUsageRecord>) {
        if (records.isEmpty()) return
        requireAuth()
        batchWrite(records, {
            val docId = "${it.date}_${it.packageName.replace(".", "_")}"
            childRef().collection("appUsage").document(docId)
        })
        Log.d(TAG, "Uploaded ${records.size} app usage records")
    }

    suspend fun updateDeviceProfile(profile: DeviceProfile) {
        requireAuth()
        childRef().collection("profile").document("info")
            .set(profile, SetOptions.merge()).await()
    }

    // ─── Download (parent commands) ──────────────────────────────────────────

    suspend fun getBlockedApps(): List<BlockedApp> {
        return try {
            val snapshot = childRef().collection("blockedApps").get().await()
            snapshot.toObjects(BlockedApp::class.java)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get blocked apps", e)
            emptyList()
        }
    }

    suspend fun getScreenTimeLimits(): ScreenTimeLimit {
        return try {
            val doc = childRef().document("screenTimeLimits").get().await()
            doc.toObject(ScreenTimeLimit::class.java) ?: ScreenTimeLimit()
        } catch (e: Exception) {
            ScreenTimeLimit()
        }
    }

    // ─── App Messages (WhatsApp, Telegram, Discord, etc.) ────────────────────

    suspend fun uploadAppMessages(messages: List<AppMessage>) {
        if (messages.isEmpty()) return
        requireAuth()
        batchWrite(messages, { childRef().collection("appMessages").document(it.id) })
        Log.d(TAG, "Uploaded ${messages.size} app messages")
    }

    // ─── Real-time listener for parent commands ───────────────────────────────

    fun listenForCommands(onBlockedAppsChanged: (List<BlockedApp>) -> Unit) {
        childRef().collection("blockedApps")
            .addSnapshotListener { snapshot, error ->
                if (error != null || snapshot == null) return@addSnapshotListener
                val apps = snapshot.toObjects(BlockedApp::class.java)
                onBlockedAppsChanged(apps)
            }
    }

    companion object {
        private const val TAG = "FirebaseRepository"
    }
}
